polygeist/tools/cgeist/Lib/Attributes.cc

```diff
@@ -7,166 +7,367 @@
+  auto Invoke = [this](AddAttrFuncPtr AddAttrPtr,
+                       auto... Args) -> AttrBuilder & {
+    return (this->*AddAttrPtr)(Args...);
+  };
+
+  NamedAttribute NamedAttr =
+      createNamedAttr(createStringAttr(AttrName, Ctx), Attr);
+  return Invoke(AddAttrPtr, NamedAttr);
 }
 
 AttrBuilder &
-AttrBuilder::addPassThroughAttribute(llvm::Attribute::AttrKind kind) {
-  StringRef attrName = llvm::Attribute::getNameFromAttrKind(kind);
-  return addPassThroughAttribute(createStringAttr(attrName));
+AttrBuilder::addPassThroughAttribute(llvm::Attribute::AttrKind Kind) {
+  return addAttribute(Kind, llvm::None, &AttrBuilder::addPassThroughAttribute);
 }
 
-AttrBuilder &AttrBuilder::addPassThroughAttribute(mlir::Attribute attr) {
-  NamedAttribute passThrough = getOrCreatePassThroughAttr();
-  assert(passThrough.getValue().isa<ArrayAttr>() &&
-         "passthrough attribute should have an ArrayAttr as value");
+AttrBuilder &
+AttrBuilder::addPassThroughAttribute(llvm::Attribute::AttrKind Kind,
+                                     mlir::Type Ty) {
+  return addAttribute(Kind, Ty, llvm::None,
+                      &AttrBuilder::addPassThroughAttribute);
+}
+
+AttrBuilder &
+AttrBuilder::addPassThroughAttribute(llvm::Attribute::AttrKind Kind,
+                                     uint64_t Val) {
+  return addAttribute(Kind, Val, &AttrBuilder::addPassThroughRawIntAttr);
+}
+
+AttrBuilder &AttrBuilder::addPassThroughAttribute(StringRef AttrName,
+                                                  mlir::Attribute Attr) {
+  return addAttribute(AttrName, Attr, &AttrBuilder::addPassThroughAttribute);
+}
+
+AttrBuilder &AttrBuilder::addPassThroughAttribute(mlir::NamedAttribute Attr) {
+  NamedAttribute PassThroughAttr = getOrCreatePassThroughAttr();
+  addToPassThroughAttr(PassThroughAttr, Attr, Ctx);
+  return addAttribute(PassThroughAttr);
+}
+
+void AttrBuilder::addToPassThroughAttr(NamedAttribute &PassThroughAttr,
+                                       mlir::NamedAttribute Attr,
+                                       MLIRContext &Ctx) {
+  assert(PassThroughAttr.getName() == PassThroughAttrName &&
+         "PassThroughAttr is not valid");
+  assert(PassThroughAttr.getValue().isa<ArrayAttr>() &&
+         "PassThroughAttr should have an ArrayAttr as value");
+
+  LLVM_DEBUG(llvm::dbgs() << "Adding attribute " << Attr.getName() << " to '"
+                          << PassThroughAttrName << "'.\n";);
 
-  LLVM_DEBUG(llvm::dbgs() << "Adding attribute " << attr << " to '"
-                          << passThroughAttrName << "'.\n");
-  std::vector<mlir::Attribute> vec =
-      passThrough.getValue().cast<ArrayAttr>().getValue().vec();
-  vec.push_back(attr);
-  passThrough.setValue(ArrayAttr::get(&ctx, vec));
+  std::vector<mlir::Attribute> Vec =
+      PassThroughAttr.getValue().cast<ArrayAttr>().getValue().vec();
+
+  // TODO: find a way to add the attributes only if one does not exist already,
+  // and keep the list in sorted order.
```

Comment:
why should it be in sorted order?

Reply:
Is not an hard requirement but it would be easier to write LIT tests if we could sort it. Anyway, this is a TODO for future improvement beyond the scope of this PR.

Reply:
oic, just wanted to know if it is implementation requirement or not.